README.md
# moxa-v2406c-hotswapd

### Files:
mxhtsp_lx.c - source of library
mxhtspd.c mxhtspd.h - source of mxhtspd daemon
mxsetled.c - source code to setup led status
mxhtsp_ioctl.h - define ioctl number and structure negotiating with driver

#### Compile the hotswap daemon
```
$ make
```

#### Clean the hotswap daemon
```
$ make clean
```


debian/changelog
moxa-v2406c-hotswapd (1.0.2+deb9) stretch; urgency=low

  * fix: plugged/unplugged disk message needs to broadcast all user

 -- Elvis Yao (姚家瑋) <ElvisCW.Yao@moxa.com>  Tue, 28 Jul 2020 11:07:29 +0800

moxa-v2406c-hotswapd (1.0.1+deb9) stretch; urgency=low

  * fix: add no-automatic-dbgsym to debian rules

 -- Elvis Yao (姚家瑋) <ElvisCW.Yao@moxa.com>  Tue, 19 Nov 2019 10:37:38 +0800

moxa-v2406c-hotswapd (1.0.0+deb9) stretch; urgency=low

  * Initial release.

 -- Elvis Yao (姚家瑋) <ElvisCW.Yao@moxa.com>  Tue, 16 Jul 2019 10:49:50 +0800


mxhtsp.h
/*
  Copyright (C) MOXA Inc. All rights reserved.
  This software is distributed under the terms of the
  MOXA License.  See the file COPYING-MOXA for details.
*/
#ifndef _MXHTSP_H_
#define _MXHTSP_H_

#ifdef __cplusplus
extern "C" {
#endif

#include "mxhtsp_ioctl.h"

#ifdef WIN32
#include <windows.h>
#endif



/**
 * Set the led to on /off
 * @param fd the open port
 * @param led_num the led number
 * @param on 1:on 0:off
 * @return  0 on success, otherwise return -1;
 */
int mxhtsp_set_led(int fd, int led_num, int on);

/**
 * Check if a button is pressed
 * @param fd the open port
 * @param btn_num the button number 
 * @return  1:pressed 0:not pressed -1:fail
 */
int mxhtsp_is_button_pressed(int fd, int btn_num);

/**
 * Check if a disk is busy
 * @param fd the open port
 * @param disk_num the disk number 
 * @return  1:busy 0:idle -1:fail
 */
int mxhtsp_is_disk_busy(int fd, int disk_num);

/**
 * Check if a disk is plugged
 * @param fd the open port
 * @param disk_num the disk number 
 * @return  1:plugged 0:unplugged -1:fail
 */
int mxhtsp_is_disk_plugged(int fd, int disk_num);

/**
 * Open the hotswap devices
 * @return	fd on success, otherwise returns -1
 */
int mxhtsp_open(void);

/**
 * Get the used percentage of a partition
 * @param partition_name the name of partition being checked. In linux, it should be /media/diskxpx
 * @return	percentage the used percentage, otherwise return -1
 */
int mxhtsp_check_partition_usage(const char *partition_name);

/**
 * Close the hotswap devies
 * @param fd the open port
 * @return  None
 */
#ifdef WIN32
void mxhtsp_close(int fd);
#else
#define mxhtsp_close(fd) close(fd)
#endif


#ifdef __cplusplus
}
#endif

#endif /* _MXHTSP_H_ */


mxhtspd.h
#define TIME_SLEEP_MS 100  /* ms */
#define TIME_BTN_LONG_PRESSED 2 /* sec, please add mlseep time */
#define TIME_PARTITION_CHECK 60 /* sec */
#define TIME_DELAY_CHECK_USAGE 10 /* delay for os to recognize partition */
#define PID_FILE "/var/run/mxhtspd.pid"

#ifdef DEBUG
	#define d(fmt, args...) printf("%s: "fmt, __FUNCTION__, ##args)
	#define dd(fmt, args...)
#else
	#define d(fmt, args...)
	#define dd(fmt, args...)
#endif


typedef struct _htsp_info {
	int disk_plugged;
	int btn_last_pressed;	/* set when the last time button is pressed */
	int btn_event_start;	/* set when button change from unpressed mode to pressed mode */
	int btn_time_accum;
	int led_on;
	int led_warn;			/* set when a partiton on disk is over usage limitation */
	int led_last_warn;		/* set when the last time led is warning */
	int first_insert;       /* Ignore btn status when disk is being insert. */
	clock_t press_time;
} htsp_info;

typedef struct _daemon_ctrl {
	int time_btn_long;
	int time_chk_part;
	int force_chk_part;		/* when disk is plugged, force to recheck */
	int log_facility;		/* set -1 if null, or 0 to 7 for local0 ~ local7 */
	int daemonize;
	int verbose;
	int exit;
} daemon_ctrl;

typedef struct _clist {
	char *mpoint;
	int percent;
	struct _clist *next;
} clist;

typedef struct _parg {
	int fd;
	htsp_info *info;
	clist *cl;
} parg;



mxhtsp_lx.c
/*  Copyright (C) MOXA Inc. All rights reserved.

    This software is distributed under the terms of the
    MOXA License.  See the file COPYING-MOXA for details.

    2010-12-10 Wade Liang Wade.Liang@moxa.com
		new release
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include "mxhtsp.h" /* prototype */

int mxhtsp_set_led(int fd, int led_num, int on)
{
	mxhtsp_param p;

	p.disk_num = led_num;
	p.val = on;
	return ioctl(fd, IOCTL_SET_DISK_LED, &p);
}

int mxhtsp_is_button_pressed(int fd, int btn_num)
{
	mxhtsp_param p;

	p.disk_num = btn_num;
	p.val = 0;
	if (ioctl(fd, IOCTL_GET_DISK_BTN, &p)) {
		return -1;
	} else {
		return p.val;
	}
}

int mxhtsp_is_disk_busy(int fd, int disk_num)
{
	mxhtsp_param p;

	p.disk_num = disk_num;
	p.val = 0;
	if (ioctl(fd, IOCTL_GET_DISK_STATUS, &p))
		return -1;
	else
		return p.val;
}

int mxhtsp_is_disk_plugged(int fd, int disk_num)
{
	mxhtsp_param p;

	p.disk_num = disk_num;
	p.val = 0;
	if (ioctl(fd, IOCTL_CHECK_DISK_PLUGGED, &p))
		return -1;
	else
		return p.val;
}

int mxhtsp_open(void)
{
	int fd;
	fd = open("/dev/hotswap", O_RDWR);
	return fd;
}

int mxhtsp_check_partition_usage(const char *mount_point)
{
	FILE *fp = NULL;
	char buf[BUFSIZ];
	char *delim = " \n";
	char cmd[50];
	char *token_percent;
	char *token_path;
	int percent;

	sprintf(cmd, "df -h %s 2>&1", mount_point);
	if (!(fp = popen(cmd, "r"))) {
		perror("can't execute popen");
		return -1;
	}

	/* skip the first time  */
	if (!fgets(buf, BUFSIZ, fp))
		goto err;

	while (fgets(buf, BUFSIZ, fp)) {
		if (!strtok(buf, delim))
			break;
		if (!strtok(NULL, delim))
			break;
		if (!strtok(NULL, delim))
			break;
		if (!strtok(NULL, delim))
			break;
		if (!(token_percent = strtok(NULL, delim)))
			break;
		if (!(token_path = strtok(NULL, delim))) {
			break;
		} else if (strcmp(token_path, mount_point) == 0) {
			sscanf(token_percent, "%d%%", &percent);
			fclose(fp);
			return percent;
		}
	}
err:
	pclose(fp);
	return -1;
}


mx_hotswapd.sh
#!/bin/bash

### BEGIN INIT INFO
# Provides:          Start mxhtspd daemon
# Required-Start:    
# Required-Stop:
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Run /usr/bin/mxhtspd if it exist
### END INIT INFO

. /lib/lsb/init-functions
PIDFILE=/var/run/mxhtspd.pid
udevadm trigger
case "$1" in

start)
	echo "Starting mxhtspd daemon..."
	#Sleep due to driver loading delay
	sleep 1
	if [ -f /var/run/mxhtspd.pid ] ; then
		echo "WARNING! mxhtspd is running, remove /var/run/mxhtspd.pid if it terminated"
		exit 0
	fi
	#Add parameter if necessary
	mxhtspd &
	;;
stop)	
	echo "Stopping mxhtspd daemon..."
	killall mxhtspd 
	;;
restart)
	echo "Restarting mxhtspd daemon..."
	killall mxhtspd &> /dev/null
	if [ -f /var/run/mxhtspd.pid ] ; then
		echo "WARNING! mxhtspd is running, remove /var/run/mxhtspd.pid if it is no longer running"
		exit 0
	fi
	mxhtspd &
        ;;
*)	
        exit 2
        ;;
esac

exit 0


mxhtspd.c
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <sys/select.h>
#include <sys/time.h>
#include <signal.h>
#include <string.h>
#include <stdarg.h>
#include <syslog.h>
#include <pthread.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <time.h>

#include "mxhtsp.h"
#include "mxhtspd.h"

/* Jared, the control should be moved the global variable for accessing safely in another sub-function */
/* default control setting */
static daemon_ctrl control = {
	.time_btn_long = TIME_BTN_LONG_PRESSED,
	.time_chk_part = TIME_PARTITION_CHECK,
	.force_chk_part = 0,
	.log_facility = -1,
	.daemonize = 0,
	.verbose = 0,
	.exit = 0
};

static daemon_ctrl *ctrl;	/* A pointer for accessing control */

static const char *dir = "/usr/sbin/";

static void msleep(unsigned long msec)
{
	struct timeval	time;
	time.tv_sec = msec / 1000;
	time.tv_usec = (msec % 1000) * 1000;
	select(1, NULL, NULL, NULL, &time);
}

static void mprintf(const char *fmt, ...)
{
	char buf[BUFSIZ];
	va_list ap;

	va_start(ap, fmt);
	vsnprintf(buf, sizeof(buf), fmt, ap);
	va_end(ap);

	if (ctrl->log_facility >= 0)
		syslog(LOG_NOTICE, "%s", buf);
	else
		printf("mxhtspd: %s", buf);
}

static void execute_script(char *cmd, int param)
{
	char name[50];

	d("%s %d\n", cmd, param);
	sprintf(name, "%s%s %d &", dir, cmd, param);
	system(name);
}

static void handle_disk_plugged(int fd, int n, htsp_info *info)
{
	int ret;
	ret = mxhtsp_is_disk_plugged(fd, n);

	if ((ret == 1) && (info[n-1].disk_plugged == 0)) {
		mprintf("Disk %d is plugged\n", n);
		if (ctrl->verbose)
			mprintf("call action-disk-plugged %d\n", n);
		execute_script("action-disk-plugged", n);
		ctrl->force_chk_part = 1;
		info[n-1].first_insert = 1;
		mxhtsp_set_led(fd, n, 1);
		info[n-1].led_on = 1;
		sleep (1);
	} else if ((ret == 0) && (info[n-1].disk_plugged == 1)) {
		mprintf("Disk %d is unplugged\n", n);
		if (ctrl->verbose)
			mprintf("call action-disk-plugged %d\n", n);
		/*
		 * Note: partition should be umounted before unplugged, and
		 * it's advicable to umount all partition in action-disk-unplugged script.
		 * Here is just a prevention for operation mistake
		 */
		execute_script("action-disk-unplugged", n);
		ctrl->force_chk_part = 1;
		mxhtsp_set_led(fd, n, 0);
		info[n-1].led_on = 0;
	}

	info[n-1].disk_plugged = ret;
}

static int handle_button_pressed(int fd, int n, htsp_info *info)
{
	int now_press;
	double time_used;
	double time_threshold = (double)ctrl->time_btn_long * 1000.0;
	clock_t now;
	/* int *ta = &info[n-1].btn_time_accum; */
	int *btn_event_start = &info[n-1].btn_event_start;
	int *btn_last_pressed = &info[n-1].btn_last_pressed;

	now_press = mxhtsp_is_button_pressed(fd, n);

	if (now_press < 0) {
		mprintf("detect button pressed error!");
		return -1;
	}

	if (now_press == 1 && (*btn_last_pressed == 0)) {
		/* record pressed time */
		info[n-1].press_time = clock();
		*btn_event_start = 1;
		*btn_last_pressed = 1;
	} else if (now_press == 0) {
		if (*btn_last_pressed == 1) {
			now = clock();
			time_used = ((double) (now - info[n-1].press_time));
			mprintf("[btn=%d, pressed time elapsed = %f]\n", n, time_used);

			/* check elapsed time over threshold */
			if (time_used > time_threshold) {
				if (*btn_event_start && info[n-1].first_insert == 0) {
					mprintf("Button %d is pressed\n", n);
					if (mxhtsp_is_disk_busy(fd, n) == 1) { /* busy */
						mprintf("Disk %d is busy.\n", n);
						if (ctrl->verbose)
							mprintf("call action-btn-pressed %d\n", n+2);
						execute_script("action-btn-pressed", n+2);
						*btn_event_start = 0;
					} else {
						mprintf("Unmount the disk %d and unplug it.\n", n);
						if (ctrl->verbose)
							mprintf("call action-btn-pressed %d\n", n);
						execute_script("action-btn-pressed", n);
						*btn_event_start = 0;
					}
				}
			}
		}

		*btn_last_pressed = 0;
		info[n-1].first_insert = 0;
	}

	return 0;
}

void sig_handler(int signo)
{
	ctrl->exit = 1;
}

static void usage(void)
{
	printf("Usage: mxhtspd [options] \n");
	/* printf("\t-i interval in seconds to check partition usage (defulat 60 secs) \n"); */
	printf("\t-l facility_num log daemon's message with LOCAL[facility_num]\n");
	printf("\t-v run in verbose mode \n");
	printf("\t-h print usage\n");
	exit(1);
}

static void parseopt(int argc, char **argv)
{
	char opt;
	int num;
	int facility[8] = {LOG_LOCAL0, LOG_LOCAL1, LOG_LOCAL2, LOG_LOCAL3,
		LOG_LOCAL4, LOG_LOCAL5, LOG_LOCAL6, LOG_LOCAL7};

	while ((opt = getopt(argc, argv, "t:i:l:dvh")) != -1) {
		switch (opt) {
		case 't':
			num = atoi(optarg);
			if (num <= 0) {
				mprintf("can't set -t  %d", ctrl->time_btn_long);
				exit(1);
			}
			ctrl->time_btn_long = num;
			break;

		case 'i':
			num = atoi(optarg);
			if (num <= 0) {
				mprintf("can't set -i %d", ctrl->time_chk_part);
				exit(1);
			}
			ctrl->time_chk_part = num;
			break;

		case 'l':
			num = atoi(optarg);
			if ((num > 7) && (num < 0)) {
				mprintf("set a wrong log facility %d\n", num);
				exit(1);
			}
			openlog("mxhtspd", LOG_CONS | LOG_PID | LOG_NDELAY, facility[num]);
			syslog(LOG_NOTICE, "Starting mxthspd daemon...");
			d("start with local %d\n", num);
			ctrl->log_facility = num;
			break;

		case 'd':
			ctrl->daemonize = 1;
			break;

		case 'v':
			ctrl->verbose = 1;
			break;

		case 'h':
			usage();
			break;

		case '?':
			usage();
		}
	}
}

static int create_pid_file(void)
{
	FILE *fp;
	pid_t pid = getpid();

	fp = fopen(PID_FILE, "w");

	if (!fp) {
		mprintf("can't open pid file\n");
		return -1;
	}

	if (fprintf(fp, "%d\n", (int)pid) <= 0) {
		return -1;
	}

	fclose(fp);
	return 0;
}

void remove_pid_file(void)
{
	if (unlink(PID_FILE))
		mprintf("can't remove pid file");
}

static void daemonize(void)
{
	pid_t pid, sid;

	/* already a daemon */
	if (getppid() == 1)
		return;

	pid = fork();

	/* error */
	if (pid < 0) {
		exit(1);
	}

	/* fork off parent */
	if (pid > 0) {
		exit(0);
	}

	/* now execute as child process */
	umask(0);

	/* get a unique session id */
	sid = setsid();
	if (sid < 0) {
		exit(1);
	}

	/* change work directory to /, prevent to lock a directory */
	if ((chdir("/")) < 0) {
		exit(1);
	}

	close(STDIN_FILENO);
	close(STDOUT_FILENO);
	close(STDERR_FILENO);
}

int main(int argc, char *argv[])
{
	int fd;
	int i;
	htsp_info info[2] = {};

	ctrl = &control;
	parseopt(argc, argv);

	/* daemonzie */
	if (ctrl->daemonize)
		daemonize();

	fd = mxhtsp_open();
	if (fd < 0) {
		mprintf("can't open device\n");
		exit(1);
	}

	/* signal handler */
	signal(SIGTERM, sig_handler);
	signal(SIGINT, sig_handler);
	signal(SIGKILL, sig_handler);

	create_pid_file();

	while (1) {

		if (ctrl->exit)
			break;

		for (i = 1; i <= 2; i++) {
			handle_disk_plugged(fd, i, info);
			handle_button_pressed(fd, i, info);
		}

		msleep(TIME_SLEEP_MS);
	}

	mprintf("Stopping...\n");
	remove_pid_file();
	closelog();
	mxhtsp_set_led(fd, 1, 0);
	mxhtsp_set_led(fd, 2, 0);
	close(fd);
	return 0;
}
